File: java/insecure_deserialization/CartPersistence.java
import jakarta.servlet.http.*;
import java.io.*;
import java.util.Base64;

public class CartPersistence extends HttpServlet {

    // VULN 1: ObjectInputStream.readObject() on user-controlled cookie bytes
    public Object restoreCartFromCookie(HttpServletRequest req) throws Exception {
        Cookie[] cookies = req.getCookies();
        for (Cookie cookie : cookies) {
            if ("cart_data".equals(cookie.getName())) {
                byte[] cartBytes = Base64.getDecoder().decode(cookie.getValue());
                ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(cartBytes));
                return ois.readObject();
            }
        }
        return null;
    }

    // VULN 2: ObjectInputStream on HTTP request body to restore saved cart
    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ClassNotFoundException {
        InputStream body = req.getInputStream();
        ObjectInputStream ois = new ObjectInputStream(body);
        Object cart = ois.readObject();
        resp.getWriter().println("Cart restored: " + cart.toString());
    }

    // VULN 3: ObjectInputStream on uploaded file (cart export/import feature)
    public Object importCartFromFile(InputStream uploadedFile) throws IOException, ClassNotFoundException {
        ObjectInputStream ois = new ObjectInputStream(uploadedFile);
        return ois.readObject();
    }
}

File: kotlin/insecure_deserialization/CartPersistence.kt
import jakarta.servlet.http.HttpServletRequest
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import java.io.ByteArrayInputStream
import java.io.ObjectInputStream
import java.util.Base64

data class CartItem(val productId: String, val quantity: Int, val price: Double)

@RestController
class CartPersistence {

    // VULN 1: ObjectInputStream deserialization from Base64 cookie - RCE via cookie manipulation
    @GetMapping("/api/cart/restore")
    fun restoreCartFromCookie(request: HttpServletRequest): Any {
        val cartCookie = request.cookies?.firstOrNull { it.name == "cart_data" }?.value
            ?: return mapOf("items" to emptyList<CartItem>())
        val bytes = Base64.getDecoder().decode(cartCookie)
        val ois = ObjectInputStream(ByteArrayInputStream(bytes))
        val cart = ois.readObject()
        ois.close()
        return cart
    }

    // VULN 2: ObjectInputStream deserialization from raw request body - cart sync endpoint
    @PostMapping("/api/cart/sync")
    fun syncCartFromRequest(request: HttpServletRequest): Map<String, Any> {
        val ois = ObjectInputStream(request.inputStream)
        val cartData = ois.readObject()
        ois.close()
        return mapOf("status" to "synced", "cart" to (cartData as? List<*> ?: emptyList<Any>()))
    }

    // VULN 3: ObjectInputStream on uploaded file - cart import feature
    @PostMapping("/api/cart/import")
    fun importCartFromFile(
        @RequestParam("cartFile") file: MultipartFile
    ): Map<String, Any> {
        val ois = ObjectInputStream(file.inputStream)
        val importedCart = ois.readObject()
        ois.close()
        return mapOf("status" to "imported", "itemCount" to (importedCart as? List<*>)?.size)
    }
}
